cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java

```diff
@@ -112,16 +112,12 @@ public Cancellable register(Discoverable discoverable) {
     try {
       CoreV1Api api = getCoreApi();
       while (true) {
-        // First try to create the service
-        if (createV1Service(api, serviceName, discoverable)) {
-          break;
-        }
-
-        // If creation failed, we update the service.
-        // To update, first need to find the service version.
         Optional<V1Service> currentService = getV1Service(api, serviceName, discoverable.getName());
         if (!currentService.isPresent()) {
-          // Loop and try to create again
+          if (createV1Service(api, serviceName, discoverable)) {
```

Comment:
I think you need to handle the race there the service was deleted after the `get` call above. In such case, the `update` will be called and it will fail with not found (currently we only handle conflict).